File: code/packages/typescript/ml-framework-core/src/autograd.ts
/**
 * # autograd.ts — reverse-mode automatic differentiation for Tensor
 *
 * Adds two pieces of machinery on top of v0.1's Tensor:
 *
 *   1. The `Function` base class.  Every differentiable op (Add, MatMul,
 *      ReLU, ...) is a `Function` subclass that defines `forward` and
 *      `backward`.  PR #3 implements the ~15 specific subclasses; here we
 *      provide the base class and one `Identity` subclass for testing.
 *
 *   2. The `backwardImpl` function — a free standalone function that
 *      implements topological sort + reverse walk on the autograd graph.
 *      `Tensor#backward` (defined in tensor.ts) delegates to it.
 *
 * ## Why backward() is a free function, not a method
 *
 * The walker needs to read every Tensor's `gradFn` and write its `grad`.
 * If it lived inside `Tensor`, we'd have either (a) a circular dependency
 * between tensor.ts and autograd.ts, or (b) a giant tensor.ts that owns
 * both storage and graph logic.  Splitting into a free function in
 * autograd.ts keeps the modules small and lets Tensor stay focused on
 * "I am a Float32Array with a shape."
 *
 * The `tensor.ts` file does have a thin `Tensor#backward(grad?)` method
 * but it's a one-liner that imports and calls `backwardImpl`.
 *
 * ## Algorithm (same as Ruby pilot + Python reference)
 *
 *   1. If `grad` is undefined, default to ones-like (PyTorch convention;
 *      strictly only valid for scalar outputs but we allow any shape).
 *   2. DFS post-order to build the topological list upstream through the
 *      `gradFn` chain.
 *   3. Walk topo list in REVERSE.  For each non-leaf, call
 *      `gradFn.backward(nodeGrad)` to get per-input grads, accumulate
 *      them into a per-parent `gradMap` keyed on object identity (via a
 *      JS `Map`, which uses reference equality for object keys).
 *   4. For each leaf with `requiresGrad`, copy/accumulate the gradient
 *      into the public `.grad` slot.  Supports repeated `backward()`
 *      without zero_grad (PyTorch convention).
 *
 * O(V + E) where V is operations, E is tensor edges.
 */

import { Tensor } from "./tensor.js";

// ---------------------------------------------------------------------------
// Function base class — extend for every differentiable op.
// ---------------------------------------------------------------------------

export abstract class Function {
  /**
   * The tensor inputs that produced the output.  `backward()` returns
   * gradients in the same order.  Set by `apply()`; subclasses should
   * not touch.
   */
  public parents: Tensor[] = [];

  /**
   * Free-form storage for whatever the subclass wants to cache in
   * `forward()` for use in `backward()` — typically the input tensors,
   * the output, or shape metadata.  Symbol-keyed `Map`-like Record so
   * keys are namespaced per-instance.
   */
  public savedForBackward: Record<string, unknown> = {};

  /**
   * The canonical entry point for invoking an op.
   *
   *   1. Instantiate the Function (so it can hold state for backward).
   *   2. Filter `inputs` to Tensors and stash them as `parents`.  Non-Tensor
   *      args (e.g. Pow's scalar exponent) pass through to `forward` but
   *      don't appear in the autograd graph — they don't have gradients.
   *   3. Run `forward(...inputs)`.
   *   4. If any Tensor input has `requiresGrad`, mark the output the same
   *      way and wire `output.gradFn = fn` so `backwardImpl` can walk us.
   *
   * The static method signature uses a generic `T extends Function` so
   * subclasses don't lose their concrete type when calling `MyOp.apply(...)`.
   */
  static apply<T extends Function>(
    this: new () => T,
    ...inputs: unknown[]
  ): Tensor {
    const fn = new this();

    fn.parents = inputs.filter((i): i is Tensor => i instanceof Tensor);

    const output = fn.forward(...inputs);

    const needsGrad = inputs.some((i) => i instanceof Tensor && i.requiresGrad);
    if (needsGrad) {
      output.requiresGrad = true;
      output.gradFn = fn;
    }
    return output;
  }

  /** Subclasses override.  Compute the forward result. */
  abstract forward(...inputs: unknown[]): Tensor;

  /**
   * Subclasses override.  Given the gradient of the loss w.r.t. this
   * Function's output, return one gradient per `parents[i]` (in the same
   * order).  Use `null` in a slot for any parent that doesn't need a
   * gradient (rare).
   */
  abstract backward(outputGrad: Tensor): (Tensor | null)[];

  /** Friendly default for inspect / console.log output. */
  toString(): string {
    return `<${this.constructor.name} parents=${this.parents.length}>`;
  }
}

// ---------------------------------------------------------------------------
// Identity — the simplest possible Function subclass.
//
// forward(x)  → fresh Tensor with the same data as x
// backward(g) → [g]   (gradient passes through unchanged)
//
// Used by the autograd test suite to exercise the apply() / backward()
// machinery without depending on op-specific math.  The real ops land
// in PR #3 (ops.ts).
// ---------------------------------------------------------------------------

export class Identity extends Function {
  forward(...inputs: unknown[]): Tensor {
    const x = inputs[0];
    if (!(x instanceof Tensor)) {
      throw new TypeError("Identity.forward expects a Tensor argument");
    }
    // Return a NEW Tensor so object identity (`y === x`) is false even
    // though `y.equals(x)` is true.  Constructors copy, so this works.
    return new Tensor(Array.from(x.data), { shape: x.shape.slice() });
  }

  backward(outputGrad: Tensor): (Tensor | null)[] {
    // d/dx(x) = 1; gradient passes through unchanged.
    return [outputGrad];
  }
}

// ---------------------------------------------------------------------------
// backwardImpl — the actual reverse-mode autodiff walker.
//
// Called from `Tensor#backward(grad?)` in tensor.ts.  Lives here so the
// algorithm code (which is the meat of autograd) sits next to the
// Function base class.
// ---------------------------------------------------------------------------

export function backwardImpl(start: Tensor, grad?: Tensor): void {
  if (!start.requiresGrad) {
    throw new Error("backward() called on a tensor that doesn't require grad");
  }

  const seed = grad ?? Tensor.onesLike(start);
  if (!shapesMatch(seed.shape, start.shape)) {
    throw new RangeError(
      `backward grad shape [${seed.shape.join(", ")}] != tensor shape [${start.shape.join(", ")}]`,
    );
  }

  // Topological order: each tensor appears AFTER its parents.  Walking in
  // reverse processes children before parents.
  const topoOrder: Tensor[] = [];
  const visited = new Set<Tensor>();   // identity-based, perfect for our needs
  function buildTopo(t: Tensor): void {
    if (visited.has(t)) return;
    visited.add(t);
    if (t.gradFn) {
      for (const p of t.gradFn.parents) buildTopo(p);
    }
    topoOrder.push(t);
  }
  buildTopo(start);

  // gradMap: Tensor → accumulated gradient.  Map uses reference equality
  // for object keys, which is exactly what we need — the same Tensor
  // reaching multiple paths (e.g. `Add(x, x)`) collapses to one entry.
  const gradMap = new Map<Tensor, Tensor>();
  gradMap.set(start, seed);

  for (let i = topoOrder.length - 1; i >= 0; i--) {
    const node = topoOrder[i]!;
    const nodeGrad = gradMap.get(node);
    if (!nodeGrad) continue;

    if (!node.gradFn) {
      // Leaf — store/accumulate into the public grad slot.
      if (!node.requiresGrad) continue;

      if (!node.grad) {
        // First time seeing this leaf — copy the accumulated grad.
        node.grad = new Tensor(Array.from(nodeGrad.data), { shape: nodeGrad.shape.slice() });
      } else {
        // Accumulate: supports repeated backward() without zero_grad
        // (caller is expected to zero between steps when training).
        node.grad = node.grad.add(nodeGrad);
      }
      continue;
    }

    // Non-leaf — ask the Function for per-input grads, distribute them.
    const inputGrads = node.gradFn.backward(nodeGrad);
    for (let j = 0; j < node.gradFn.parents.length; j++) {
      const parent = node.gradFn.parents[j]!;
      const inputGrad = inputGrads[j];
      if (!inputGrad) continue;

      const existing = gradMap.get(parent);
      gradMap.set(parent, existing ? existing.add(inputGrad) : inputGrad);
    }
  }
}

function shapesMatch(a: readonly number[], b: readonly number[]): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

File: code/packages/ruby/ml_framework_core/lib/coding_adventures/ml_framework_core/autograd.rb
# frozen_string_literal: true

# autograd.rb — reverse-mode automatic differentiation for Tensor
# =================================================================
#
# This file adds two pieces of machinery to ml_framework_core:
#
#   1. The `Function` base class.  Every differentiable op (Add, MatMul,
#      ReLU, ...) is a `Function` subclass that defines `forward` and
#      `backward`.  PR #6 implements the ~15 specific subclasses; here
#      we just provide the base class and one Identity subclass for
#      testing the machinery.
#
#   2. `Tensor#backward(grad = nil)` — kicks off backpropagation from a
#      tensor (typically a scalar loss).  Walks the autograd graph in
#      reverse topological order, calling each Function's `backward` to
#      compute input gradients and accumulating them into leaf tensors'
#      `grad` slot.
#
# # Why it's so small
#
# All the heavy lifting in autograd is per-op math (PR #6/#7).  The
# *framework* itself — apply() plumbing + topo sort + reverse walk — is
# under 100 lines.  This is a property of well-designed reverse-mode AD:
# every paper that ever described autograd shows this exact algorithm.
#
# # Mirrors the Python structure
#
# This file deliberately mirrors
# `code/packages/python/ml-framework-core/src/ml_framework_core/autograd.py`
# in:
#   - Function base class shape (saved_tensors + saved_metadata)
#   - apply()'s steps: probe requires_grad → run forward → wire grad_fn
#   - backward()'s topo-sort + reverse-walk algorithm
#   - gradient accumulation rule (sum across paths into a node)
#
# Differences are purely Ruby idiom:
#   - `id()` → `object_id`
#   - `dict` → `Hash`
#   - `set` → `Set` (or just keying a Hash on object_id)
#   - module function `backward(tensor)` → `Tensor#backward` method

require "set"
require_relative "tensor"

module CodingAdventures
  module MLFrameworkCore
    # ========================================================================
    # Reopen Tensor — autograd needs setters for grad / grad_fn and a
    # `Tensor.ones_like` convenience factory.  Keeping them here (instead
    # of in tensor.rb) keeps the v0.1 Tensor PR strictly storage-only,
    # and concentrates the autograd-related additions in one file for
    # easy review.
    # ========================================================================
    class Tensor
      # Setters for the autograd-only attributes.  Public so Function.apply
      # and the backward() walker can write them; users shouldn't need to.
      attr_writer :grad, :grad_fn

      # `Tensor.ones_like(t)` — a tensor of 1.0s with the same shape as `t`.
      # The PyTorch convention for `backward()` with no explicit gradient.
      def self.ones_like(other)
        ones(*other.shape)
      end

      # `Tensor.zeros_like(t)` — companion to ones_like; used here for the
      # leaf-grad initialization branch in backward().
      def self.zeros_like(other)
        zeros(*other.shape)
      end

      # Tensor#backward(grad = nil) — kick off backprop.
      #
      # Algorithm:
      #   1. If grad is nil, default to ones-like (PyTorch convention; valid
      #      strictly only for scalar tensors but we allow any shape here
      #      for simplicity, matching what PR #6's tests will need).
      #   2. Topological sort upward through grad_fn chain — DFS, post-order.
      #   3. Walk topo list in reverse; for each tensor with a grad_fn,
      #      call grad_fn.backward(tensor.grad), get input grads, accumulate
      #      into each parent's grad slot via the grad_map.
      #   4. For leaf tensors with requires_grad, copy the accumulated
      #      gradient into the tensor's `grad` slot (accumulating if a
      #      previous backward already wrote there — supports repeated
      #      backward() calls without zero_grad in between).
      #
      # Returns nil (mutates in place; matches PyTorch).
      def backward(grad = nil)
        unless requires_grad
          raise RuntimeError, "backward() called on a tensor that doesn't require grad"
        end

        grad ||= Tensor.ones_like(self)
        unless grad.shape == shape
          raise ArgumentError,
                "backward grad shape #{grad.shape.inspect} != tensor shape #{shape.inspect}"
        end

        # Topological order: each tensor appears AFTER its parents in
        # the list, so walking in REVERSE processes children before parents.
        topo_order = []
        visited    = Set.new
        build_topo = lambda do |t|
          # Visiting set keyed by object_id so duplicate sub-graphs (e.g.
          # `x + x`) are handled correctly — visit each Tensor exactly once.
          oid = t.object_id
          next if visited.include?(oid)

          visited.add(oid)
          if t.grad_fn
            t.grad_fn.parents.each { |p| build_topo.call(p) }
          end
          topo_order << t
        end
        build_topo.call(self)

        # grad_map: tensor.object_id → accumulated gradient Tensor.
        # Starts with our seed gradient at the root.
        grad_map = { object_id => grad }

        topo_order.reverse_each do |node|
          node_grad = grad_map[node.object_id]
          next if node_grad.nil?

          if node.grad_fn.nil?
            # Leaf tensor — store/accumulate into its public grad slot.
            next unless node.requires_grad

            if node.grad.nil?
              node.grad = Tensor.new(node_grad.to_a, shape: node_grad.shape)
            else
              # Accumulate: support repeated backward() calls (e.g. when
              # training a model and not calling zero_grad between steps).
              node.grad = node.grad + node_grad
            end
            next
          end

          # Non-leaf — ask the Function for input grads and distribute them
          # into the grad_map for the parents.
          input_grads = node.grad_fn.backward(node_grad)
          input_grads = [input_grads] unless input_grads.is_a?(Array)
          node.grad_fn.parents.each_with_index do |parent, i|
            input_grad = input_grads[i]
            next if input_grad.nil?

            existing = grad_map[parent.object_id]
            grad_map[parent.object_id] =
              if existing.nil?
                input_grad
              else
                # Two paths converged on this parent — sum the gradients.
                # (E.g. in `x + x`, the same `x` reaches the add twice.)
                existing + input_grad
              end
          end
        end

        nil
      end
    end

    # ========================================================================
    # The Function base class — extend this for every differentiable op.
    # ========================================================================
    class Function
      # @return [Array<Tensor>] the input tensors that produced the output.
      #   backward() distributes gradients to these in the same order.
      attr_accessor :parents

      # @return [Hash{Symbol => Object}] subclass-defined values cached in
      #   forward() for use in backward() — typically input shapes, the
      #   activation values needed for derivative computation (e.g.
      #   sigmoid backward needs the output value), etc.
      attr_accessor :saved_for_backward

      def initialize
        @parents = []
        @saved_for_backward = {}
      end

      # Class method — the canonical entry point for invoking an op.
      #
      # Algorithm:
      #   1. Instantiate the Function (so it can hold state for backward).
      #   2. Call its `forward(*inputs)` to compute the output Tensor.
      #   3. If any input has `requires_grad`, mark the output the same
      #      way and wire `output.grad_fn = self` so `Tensor#backward` can
      #      find us later.
      #
      # Returns the output Tensor.
      def self.apply(*inputs)
        fn = new

        # Stash the parent TENSORS so backward() can distribute grads
        # back to them.  Non-Tensor args (e.g. Pow's scalar exponent)
        # are dropped here — they don't participate in autograd and
        # `build_topo` in Tensor#backward would crash trying to read
        # `grad_fn` on a non-Tensor.  Subclasses that need a non-Tensor
        # arg can stash it in @saved_for_backward.
        #
        # Note: we save the actual Tensor objects; object identity
        # matters for shared-parameter scenarios like `Add(x, x)`
        # where the same Tensor appears twice — backward needs to
        # accumulate both gradient contributions into the same `.grad`.
        fn.parents = inputs.select { |i| i.is_a?(Tensor) }

        output = fn.forward(*inputs)

        # If ANY input tracks gradients, the output tracks them too — and
        # we record this Function as the source of the output, so backward
        # can walk back through us.
        needs_grad = inputs.any? { |t| t.is_a?(Tensor) && t.requires_grad }
        if needs_grad
          output.requires_grad = true
          output.grad_fn = fn
        end

        output
      end

      # Subclasses override this.  Compute the forward result.
      # @return [Tensor]
      def forward(*_inputs)
        raise NotImplementedError, "#{self.class} must implement #forward"
      end

      # Subclasses override this.  Given the gradient of the loss w.r.t.
      # this Function's output, return an Array of gradients w.r.t. each
      # input tensor (in the same order as `parents`).  Return `nil` in
      # a slot for any input that doesn't need a gradient.
      # @return [Array<Tensor, nil>]
      def backward(_output_grad)
        raise NotImplementedError, "#{self.class} must implement #backward"
      end

      # Friendly default for pp / inspect.
      def inspect
        "#<#{self.class.name} parents=#{@parents.length}>"
      end
    end

    # ========================================================================
    # Identity — the simplest possible Function subclass.
    #
    # forward(x)  → x' where x' is a fresh Tensor with the same data
    # backward(g) → [g]   (the upstream gradient passes through unchanged)
    #
    # Used by the autograd test suite to exercise the apply() / backward()
    # machinery without bringing in any op-specific math.  The real ops
    # land in PR #6 (ops.rb).
    # ========================================================================
    class Identity < Function
      def forward(x)
        # Return a NEW Tensor wrapping a copy of x's data, so the
        # autograd graph has a distinct node — pointer-equality
        # (`y.equal?(x)`) is false even though `y == x` is true.
        Tensor.new(x.to_a, shape: x.shape, dtype: x.dtype)
      end

      def backward(output_grad)
        # Identity's derivative is 1, so the gradient passes through
        # unchanged.  Wrap as an Array so apply() can distribute it to
        # parents[0] uniformly.
        [output_grad]
      end
    end
  end
end
